palace/fem/qfunctions/utils_qf.h

```diff
@@ -0,0 +1,522 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef PALACE_LIBCEED_UTILS_QF_H
+#define PALACE_LIBCEED_UTILS_QF_H
+
+#include <math.h>
+
+CEED_QFUNCTION_HELPER CeedScalar DetJ22(const CeedScalar *J, const CeedInt J_stride)
+{
+  // J: 0 2
+  //    1 3
+  return J[J_stride * 0] * J[J_stride * 3] - J[J_stride * 1] * J[J_stride * 2];
+}
+
+CEED_QFUNCTION_HELPER CeedScalar DetJ21(const CeedScalar *J, const CeedInt J_stride)
+{
+  // J: 0
+  //    1
+  return sqrt(J[J_stride * 0] * J[J_stride * 0] + J[J_stride * 1] * J[J_stride * 1]);
+}
+
+CEED_QFUNCTION_HELPER CeedScalar DetJ33(const CeedScalar *J, const CeedInt J_stride)
+{
+  // J: 0 3 6
+  //    1 4 7
+  //    2 5 8
+  return J[J_stride * 0] *
+             (J[J_stride * 4] * J[J_stride * 8] - J[J_stride * 5] * J[J_stride * 7]) -
+         J[J_stride * 1] *
+             (J[J_stride * 3] * J[J_stride * 8] - J[J_stride * 5] * J[J_stride * 6]) +
+         J[J_stride * 2] *
+             (J[J_stride * 3] * J[J_stride * 7] - J[J_stride * 4] * J[J_stride * 6]);
+}
+
+CEED_QFUNCTION_HELPER CeedScalar DetJ32(const CeedScalar *J, const CeedInt J_stride)
+{
+  // J: 0 3
+  //    1 4
+  //    2 5
+  const CeedScalar E = J[J_stride * 0] * J[J_stride * 0] +
+                       J[J_stride * 1] * J[J_stride * 1] +
+                       J[J_stride * 2] * J[J_stride * 2];
+  const CeedScalar G = J[J_stride * 3] * J[J_stride * 3] +
+                       J[J_stride * 4] * J[J_stride * 4] +
+                       J[J_stride * 5] * J[J_stride * 5];
+  const CeedScalar F = J[J_stride * 0] * J[J_stride * 3] +
+                       J[J_stride * 1] * J[J_stride * 4] +
+                       J[J_stride * 2] * J[J_stride * 5];
+  return sqrt(E * G - F * F);
+}
+
+CEED_QFUNCTION_HELPER void MultAdjJCAdjJt22(const CeedScalar *J, const CeedInt J_stride,
+                                            const CeedScalar *c, const CeedInt c_stride,
+                                            const CeedInt c_comp, const CeedScalar qw,
+                                            const CeedInt qd_stride, CeedScalar *qd)
+{
+  // Compute qw / det(J) adj(J) C adj(J)^T and store the symmetric part of the result.
+  // J: 0 2   adj(J):  J22 -J12   qd: 0 1
+  //    1 3           -J21  J11       1 2
+  const CeedScalar J11 = J[J_stride * 0];
```

Comment:
The 1-based indexing for variable names is just because typically it is how one refers to entries of a matrix. That's the only thing it's used for, obviously the other approach would work just as well.